Implementation of GH checks

lwintjen opened the pull request:
I added two scenarios tests called _githubflow-automerge-githubchecks-standard_ and _githubflow-automerge-githubchecks-all-checks-successful_ which have almost the same logic as _githubflow-nonbatched-all-checks-successful_ but to know if the test passed or failed, it tests existing check(s) on a PR (by querying the checks API) : 

1. at least one check should contain _maestro-policy_ in his external ID 
2. every check with an external ID containing  _maestro-policy_ should be in a completed state

If the two points above are OK then the test passes, otherwise it fails.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/arcade-services?pullRequest=1301) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/arcade-services?pullRequest=1301) before we can accept your contribution.<br/><br/>:x: lwintjen [sign now](https://cla.dotnetfoundation.org/dotnet/arcade-services?pullRequest=1301)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/arcade-services?pullRequest=1301) it.</sub>

missymessa commented:
Hi @lwintjen! We're actually converting these tests to C# and moving away from the Powershell scenario teats. Can you touch base with @meganquinn about how to add your tests to the C# project? Thanks!

lwintjen commented:
Hey @missymessa , thanks for the information. I'll reach out to @meganaquinn

lwintjen commented:
The scenario test runs successfully with a merge policy set to standard (f.e. [check this PR](https://github.com/maestro-auth-test/maestro-test2/pull/11240)). I still need to investigate what happens if the merge policy is all-checks-successful. That'd probably require some changes in _EvaluateChecksAsync_ method (AllChecksSuccessfulMergePolicy.cs) to ignore the checks with an externalID set to _maestro-policy-{}-{}_

As for the conversion of the scenario test from Powershell to C#, this will be done next week as the [PR](https://github.com/dotnet/arcade-services/pull/1313) from @meganaquinn  get merged.

This is the first version of the implementation, it probably requires some refactoring  and more features will be added to it.

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -417,6 +418,110 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+                // If the check exist, checks that the status are different to update it
+                else if (existingCheckRun != null && newCheckRun.Status != existingCheckRun.Status)
+                {
+                    CheckRunUpdate updatedCheck = new CheckRunUpdate();
+                    updatedCheck.Status = newCheckRun.Status;
+                    updatedCheck.Conclusion = newCheckRun.Conclusion;
+                    updatedCheck.Name = newCheckRun.Name;
+                    updatedCheck.CompletedAt = newCheckRun.CompletedAt;
+                    existingChecksList.Remove(existingCheckRun);
```
Please create method ToCheckRunUpdate(newCheckRun )

rokonec replied:
removing from list should happen even if status do not change as result of existingChecksList after this loop shall not contain ANY actual checks from evaluation

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -417,6 +418,110 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+            {
+                MergePolicyEvaluationResult.SingleResult remainingCheckCommon = evaluations.SingleOrDefault(eval => remainingCheck.ExternalId == $"maestro-policy-{eval.MergePolicyName}-{latestCommit.Sha}");
+                // Avoid deleting check(s) that aren't from maestro
+                if (remainingCheckCommon != null)
```
remainingCheck should never be in evaluations, so this will always be false. Consider to use  `if (remainingCheck.ExternalId.StartsWith("maestro-policy-"))`

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -417,6 +418,110 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+                return newCheck;
+
+            }
+            var newCheckRun = new NewCheckRun($"{result.MergePolicyDisplayName} {result.Message}", sha);
```
newCheckRun should always have name of result.MergePolicyDisplayName - without result.Message.
This will result into `var newCheck = new NewCheckRun($"{result.MergePolicyDisplayName}", sha);` be same in all cases and hence could be removed from these cases and moved above in higher scope. (example line 476)

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -417,6 +418,110 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+            newCheckRun.Output = newCheckRunOutput;
+            newCheckRun.ExternalId = $"maestro-policy-{result.MergePolicyName}-{sha}";
+            newCheckRun.Status = CheckStatus.Completed;
+            newCheckRun.Conclusion = "success";
```
Bug, should be .Failed

lwintjen commented on src/Microsoft.DotNet.Darc/src/DarcLib/IGitRepo.cs:
```diff
@@ -2,6 +2,8 @@
 // See the LICENSE file in the project root for more information.
 
+using Maestro.Contracts;
+using Octokit;
```
Octokit is indeed unused but Maestro.Contract is used in the method CreateOrUpdatePullRequestMergeStatusInfoAsync parameters as it uses MergePolicyEvaluationResult data model. 

`  Task CreateOrUpdatePullRequestMergeStatusInfoAsync(string pullRequestUrl,IReadOnlyList<MergePolicyEvaluationResult.SingleResult> evaluations);`

rokonec replied:
Both using Octokit; and using System.Net.Http; shall be deleted

rokonec commented on src/Maestro/tests/Maestro.ScenarioTests/Properties/serviceDependencies.json:
```diff
@@ -0,0 +1,7 @@
+{
```
I believe this file shall not be tracked. Please delete it

rokonec commented on src/Maestro/tests/Maestro.ScenarioTests/Properties/serviceDependencies.local.json:
```diff
@@ -0,0 +1,7 @@
+{
```
I believe this file shall not be tracked. Please delete it

rokonec commented on src/Maestro/tests/Maestro.ScenarioTests/TestParameters.cs:
```diff
@@ -49,7 +49,7 @@ public static async Task<TestParameters> GetAsync()
             await TestHelpers.RunExecutableAsync(dotnetExe, toolInstallArgs.ToArray());
 
-            string darcExe = Path.Join(testDir.Peek()!.Directory, RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "darc.exe" : "darc");
+            string darcExe = "darc.exe"; //Path.Join(testDir.Peek()!.Directory, RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "darc.exe" : "darc");
```
Please do not forget to revert it back.

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -417,6 +418,95 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+            // Get a list of all the merge policies checks runs for the current PR
+            List <CheckRun> existingChecksRuns = (await Client.Check.Run.GetAllForReference(owner, repo, prSha)).CheckRuns.Where(e => e.ExternalId.StartsWith("maestro-policy-")).ToList(); 
+
+            var toBeAdded = evaluations.Where(e => existingChecksRuns.All(c => c.ExternalId != $"maestro-policy-{e.MergePolicyName}-{prSha}"));
```
```suggestion
            var toBeAdded = evaluations.Where(e => existingChecksRuns.All(c => c.ExternalId != CheckRunId(e, prSha)"));
```

rokonec replied:
Please create CheckRunId(or other named) and use it everywhere when $"maestro-policy-{e.MergePolicyName}-{prSha}" is currently used

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -476,45 +476,101 @@ private NewCheckRun CheckRunForAdd(MergePolicyEvaluationResult.SingleResult resu
+        {
+            CheckRunUpdate updatedCheckRun = new CheckRunUpdate();
+            updatedCheckRun.Name = checkRun.Name;
+            updatedCheckRun.ExternalId = checkRun.ExternalId;
```
In this case you don't need to set properties which are not gonna change. I believe that null values from CheckRunUpdate are not changed. Can you please validate it on your prove of concept?

rokonec commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -476,45 +476,101 @@ private NewCheckRun CheckRunForAdd(MergePolicyEvaluationResult.SingleResult resu
+            }
+
+            CheckRunUpdate updatedCheckRun = new CheckRunUpdate();
+            updatedCheckRun.Name = checkRun.Name;
```
In this case you don't need to set properties which are not gonna change. I believe that null values from CheckRunUpdate are not changed. Can you please validate it on your prove of concept?

lwintjen replied:
Yes, you're right. I changed that (by removing the part where I set name and external ID when updating or deleting a check run). 
Thanks.

rokonec commented on src/Maestro/tests/Scenarios/githubflow-automerge-githubchecks-all-checks-successful.ps1:
```diff
@@ -0,0 +1,87 @@
+param(
```
please delete this file. moved into c#

lwintjen commented:
I finally implemented the Checks API logic into Maestro and converted the powershell scenario tests to C#  (which test the implementation of the Checks API via).

The scenario tests in C# pass, everything seems to work & I already refactored the code. I think  the PR is ready to be reviewed.

mmitche commented:
Awesome! Will take a look.

ChadNedzlek commented on src/Maestro/Maestro.Contracts/MergePolicyEvalutationResult.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public bool Succeeded => Results.Count > 0 && Results.All(r => r.Status == MergePolicyEvaluationStatus.Success);
+
+        public bool Pending => Results.Count > 0 && Results.Any(r => r.Status == MergePolicyEvaluationStatus.Pending);
```
Any already returns false if the thing is empty, so the "Results.Count" is unnecessary. (Weirdly the same is _not_ true of All, so it is required there)

ChadNedzlek commented on src/Maestro/Maestro.Contracts/MergePolicyEvalutationResult.cs:
```diff
@@ -0,0 +1,57 @@
+            Results = results.ToImmutableList();
+        }
+
+        public IReadOnlyList<MergePolicyEvaluationResult> Results { get; }
```
I wonder if returning IImmutalbeList is better. It's most explicit about the contract. (Just because something is IReadOnly doesn't mean it won't change, so it's not really safe to assume it won't.  IImmutable is a much "stronger" contract that lets the caller do a lot more safely)

ChadNedzlek commented on src/Maestro/Maestro.Contracts/MergePolicyEvalutationResult.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            if (mergePolicy.Name == null)
+            {
+                throw new ArgumentNullException(nameof(mergePolicy.Name));
```
This is going to produce a weird exception, because it's just going to say that the "Name" argument is null... but this method doesn't have a "Name" parameter.  It should probably just be an ArgumentException with some text that explains it.

It's also odd that they _actual_ argument, "mergePolicy" isn't checked, and will just crash with a NullReferenceException.

ChadNedzlek commented on src/Maestro/Maestro.MergePolicies/MergePolicy.cs:
```diff
@@ -47,6 +49,41 @@ public string Name
-        public abstract Task EvaluateAsync(IMergePolicyEvaluationContext context, MergePolicyProperties properties);
+        public abstract Task<MergePolicyEvaluationResult> EvaluateAsync(IPullRequest pr, IRemote darc);
+
+        public Task<MergePolicyEvaluationResult> Pending(string message)
```
Why are these async? They aren't virtual, and there is no chance they will do anything Async.

rokonec replied:
I have decided to have this async for two reasons:
1. enable of adding some async into these without changing calling code. For example writing into db.
1. IMergePolicy.EvaluateAsync are sometimes implemented without await in body. In such situation there would have to be `return Task.FromResult(Fail("there was an error"))`. I always optimize code for readability and Task.FromResult lover readability. So I chosed to have lower readability in helper methods, that in core logic methods. This ways IMergePolicy.EvaluateAsync can use `return Fail("error")` or `return await Fail("error")` - both of those perfectly readable

ChadNedzlek replied:
We can change code later... making methods return Task just in case is a little weird...  I can't imagine a world where `Fail("Some message")` is async... what sort of scenarios are you envisioning?

rokonec replied:
Possible future cases of await in Fail(""):
- writing results into storage
- sending telemetry data by event hub

But actually the 2 point - increased readability of all MergePolicy.EvaluateAsync implementation - is the main driver here.

ChadNedzlek commented on src/Maestro/Maestro.MergePolicies/MergePolicy.cs:
```diff
@@ -47,6 +49,41 @@ public string Name
+
+    public interface IMergePolicyBuilder
+    {
+        public string Name { get; }
```
This isn't valid code... the fact that the build completed is... troublesome.  Interface members can't have access modifiers...

ChadNedzlek replied:
Even if it is, it's... highly unusual (GitHub doesn't even understand enough to highlight it), so we should probably remove it.

rokonec replied:
Good catch. R# would tell, but ... :-(

ChadNedzlek commented on src/Maestro/Maestro.MergePolicies/MergePolicy.cs:
```diff
@@ -47,6 +49,41 @@ public string Name
+        /// In most cases it will return array of exactly one merge policy, but in special cases like standard-policy
+        /// it will return multiple pre-configured policies which that policies template consist of
+        /// </summary>
+        /// <param name="properties"></param>
```
Remove the param/returns comments. The param comment list isn't complete, and there is nothing useful in the elements anyway.  Bad/empty comments are worse than no comments.

ChadNedzlek commented on src/Maestro/Maestro.MergePolicies/StandardMergePolicy.cs:
```diff
@@ -2,55 +2,73 @@
-            standardAzureDevOpsIgnoredChecks = new HashSet<string>
+                { 
+                    "ignoreChecks", 
+                    JToken.FromObject(new []
```
I think this can be `new JArray("WIP", "license/cla", "auto-merge.config.enfoce")`, which avoids the new anonymous type and the need to serialize it.

rokonec replied:
new [] {..} it is not anonymous type but shortcut for new string[] {...}, but you are right new JArray is better here

ChadNedzlek commented on src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:
```diff
@@ -431,5 +438,52 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
         {
             return await RunDarcAsync("get-repository-policies", "--all", "--repo", repoUri, "--branch", branchName);
         }
+        public async Task<bool> WaitForMergedPullRequestAsync(string targetRepo, string targetBranch, PullRequest pr, Repository repo, int attempts = 7)
```
It looks likes this method can only return "true", so it can probably just be "Task"

ChadNedzlek commented on src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:
```diff
@@ -431,5 +438,52 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
+        public async Task<bool> ValidateGithubMaestroCheckRunsSuccessful(string targetRepoName, string targetBranch, PullRequest pullRequest, Repository repo)
+        {
+            // Waiting 5 minutes for maestro to add the checks to the PR
+            await Task.Delay(300000);
```
await Task.Delay(TimeSpan.FromMinutes(5)) is a lot easier to not typo (I know I'm the worst for having one too many or two few zeros), and to code review.

ChadNedzlek commented on src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_MergePolicies.cs:
```diff
@@ -0,0 +1,119 @@
+            var targetRepo = "maestro-test2";
+            var targetBranch = _random.Next(int.MaxValue).ToString();
+
+            TestContext.WriteLine("GitHub Dependency Flow, non-batched, no requested changes merge policies");
```
This looks like it's just repeating the test name?

ChadNedzlek commented on src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_MergePolicies.cs:
```diff
@@ -0,0 +1,119 @@
+    public class ScenarioTests_MergePolicies : MaestroScenarioTestBase
+    {
+        private TestParameters _parameters;
+        private Random _random = new Random();
```
Random is really scary in tests, because it means they can randomly fail.  What sorts of stuff are we trying to accomplish with random, and is there some other way to accomplish it?

lwintjen replied:
Create a random ID for the channel name. It was the way it was used in the powershell script (by using a random function).

ChadNedzlek commented on src/Maestro/tests/Maestro.ScenarioTests/TestParameters.cs:
```diff
@@ -21,7 +21,7 @@ public static async Task<TestParameters> GetAsync()
                 .Build();
 
-            string maestroBaseUri = Environment.GetEnvironmentVariable("MAESTRO_BASEURI") ?? "https://maestro-int.westus2.cloudapp.azure.com";
+            string maestroBaseUri = Environment.GetEnvironmentVariable("MAESTRO_BASEURI") ??  userSecrets["MAESTRO_BASEURI"] ?? "https://maestro-int.westus2.cloudapp.azure.com";
```
The BaseURI isn't really a secret. Why are we reading it out of the userSecrets dictionary?

lwintjen replied:
To avoid setting the BaseURI as a env variable every time you re-run ngrok. It's much more easier to change the userSecrets but I'm not the one that actually did that change so Roman might have a better answer to this.

rokonec replied:
Tu run local scenarios you have to run ngrok which generates random  url for BaseURI. In the former code you would have to set env variables and consequently restart visual studio to pick up env changes.
Editing secret json file is better experience, you can do it from visual studio and the changes are applied without restart of visual studio.
Also actually, ngrok URI is kinda secret as attacker could use it to attack your local running services. Luckily ngrok session has limited runtime, half a day max, so the attack window is small and actual risk negligable. But still :-)

ChadNedzlek commented on src/Microsoft.DotNet.Darc/src/Darc/Constants.cs:
```diff
@@ -51,35 +46,35 @@ public class Constants
             "See below for available merge policies:",
             "",
-            $"{StandardMergePolicyName} - Corresponds to either StandardGitHub or StandardAzureDevOps depending on the target repo type",
+            $"{MergePolicyConstants.StandardMergePolicyName} - Corresponds to either StandardGitHub or StandardAzureDevOps depending on the target repo type",
```
I really appreciate these constants getting broken up into more reasonably encapsulated types. :-)  No change... just wanted to say I was happy. :-)

ChadNedzlek commented on src/Microsoft.DotNet.Darc/src/DarcLib.AzDev/AzureDevOpsClient.cs:
```diff
@@ -527,6 +528,32 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+        {
+            await CreateOrUpdatePullRequestCommentAsync(pullRequestUrl,
+                    $@"## Auto-Merge Status
+                    This pull request has not been merged because Maestro++ is waiting on the following merge policies.
```
Do we really want this giant indent in the comment?  Any leading space in a verbatim string is included verbatim in the value.

lwintjen replied:
We can change that, I kept the same values as it was before the implementation of the checks.

mmitche replied:
I wonder if the the markdown renderer just removes the extra spaces.

ChadNedzlek replied:
Leading spaces mark the text as a code block, so it will get <pre> rendered in a big gray box.

mmitche replied:
##This has no leading spaces
    but this does.

mmitche replied:
Okay yeah. Initial heading and then code block.

It looks like the original code removed the leading spaces on lines 2+, so I'd keep it that way.

mmitche replied:
I think this still needs to be fixed, but otherwise looks good.

mmitche commented on src/Maestro/Maestro.Contracts/MergePolicyConstants.cs:
```diff
@@ -0,0 +1,21 @@
+using System;
```
Nit: License header

mmitche commented on src/Maestro/Maestro.Contracts/IMergePolicy.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Maestro.Contracts
+{
+    public interface IMergePolicyInfo
```
Nit: the file name should match the class name here.

mmitche commented on src/Maestro/SubscriptionActorService/MergePolicyEvaluator.cs:
```diff
@@ -18,40 +18,59 @@ public class MergePolicyEvaluator : IMergePolicyEvaluator
+        public MergePolicyEvaluator(IEnumerable<IMergePolicyBuilder> mergePolicies, OperationManager operations, ILogger<MergePolicyEvaluator> logger)
         {
-            MergePolicies = mergePolicies.ToImmutableDictionary(p => p.Name);
+            MergePolicieBuilders = mergePolicies.ToImmutableDictionary(p => p.Name);
```
Nit: typo (MergePolicieBuilders -> MergePolicyBuilders)

mmitche commented on src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:
```diff
@@ -105,7 +106,7 @@ public Task<string> RunDarcAsync(params string[] args)
             return TestHelpers.RunExecutableAsync(_parameters.DarcExePath, args.Concat(new[]
             {
-                "-p", _parameters.MaestroToken,
+                "--p", _parameters.MaestroToken,
```
I'm kind of surprised this syntax works (I checked). Oftenpassing -- to prefix a shortened version of a parameter is incorrect (e.g. git)

mmitche commented on src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs:
```diff
@@ -417,6 +418,152 @@ public async Task CreateOrUpdatePullRequestCommentAsync(string pullRequestUrl, s
+
+            var toBeAdded = evaluations.Where(e => existingChecksRuns.All(c => c.ExternalId != CheckRunId(e, prSha)));
+            var toBeUpdated = existingChecksRuns.Where(c => evaluations.Any(e => c.ExternalId == CheckRunId(e, prSha)));
+            var toBeDeleted = existingChecksRuns.Where(c => evaluations.All(e => c.ExternalId != CheckRunId(e, prSha)));
```
In what cases would we delete existing checks?

rokonec replied:
We were not sure if it is even possible but I though about following use case:
1. Subscription merge policy is set to standard
2. Merge is blocked because of requested-changes policy
3. Subscription merge policy is **changed** to all-checks 
4. Now there are two checks (requested-changes and downgrade) which are not in play anymore

Is this scenario valid? Are changes in subs merge policy applied to running existing pull requests?

rokonec replied:
All in all, I believe that this code has to be prepared for possibility of outdated existing checks as we can't rule this future possibility out.

mmitche replied:
I believe changes in a subscription's merge policy would apply to the current subscription. This is almost certainly true for a batched subscription.

This is a good case to cover.
